File: ingestion/chunker.py
"""Adaptive hierarchical chunking.

A section under the token threshold is stored as one whole chunk. A larger
section is split into per-sub-section child chunks that each carry a
``parent_section_id``; sub-section text is stored once (parent expansion is a
query-time concern, out of scope for ingestion).
"""
from __future__ import annotations

from typing import List

from ingestion.models import (
    AmendmentHistory,
    Chunk,
    ProvenanceRecord,
)
from ingestion.parser import ParsedAct, Section


def estimate_tokens(text: str) -> int:
    """Cheap, monotonic token estimate (whitespace words)."""
    return len(text.split())


def section_id(act_id: str, section_number: str) -> str:
    return f"{act_id}-{section_number}"


def _provenance(act: ParsedAct, section: Section, verbatim: str, sub_section=None):
    return ProvenanceRecord(
        act_name=act.act_name,
        act_year=act.act_year,
        act_type=act.act_type,
        source_url=act.source_url,
        source_hash=act.source_hash,
        retrieval_date=act.retrieval_date,
        verbatim_text=verbatim,
        section_number=section.section_number,
        sub_section=sub_section,
    )


def _amendment_history(section: Section) -> AmendmentHistory:
    return AmendmentHistory(
        entries=list(section.amendments),
        none_recorded=not section.amendments,
    )


def chunk_section(act: ParsedAct, section: Section, token_threshold: int) -> List[Chunk]:
    parent_id = section_id(act.act_id, section.section_number)
    whole = section.full_text
    if estimate_tokens(whole) <= token_threshold or not section.sub_sections:
        return [
            Chunk(
                chunk_id=parent_id,
                act_id=act.act_id,
                section_number=section.section_number,
                text=whole,
                provenance=_provenance(act, section, whole),
                amendment_history=_amendment_history(section),
                is_definition=section.is_definition,
                token_estimate=estimate_tokens(whole),
            )
        ]

    children: List[Chunk] = []
    for sub in section.sub_sections:
        children.append(
            Chunk(
                chunk_id=f"{parent_id}-{sub.label}",
                act_id=act.act_id,
                section_number=section.section_number,
                sub_section=sub.label,
                parent_section_id=parent_id,
                text=sub.text,
                provenance=_provenance(act, section, sub.text, sub_section=sub.label),
                amendment_history=_amendment_history(section),
                is_definition=section.is_definition,
                token_estimate=estimate_tokens(sub.text),
            )
        )
    return children


def chunk_act(act: ParsedAct, token_threshold: int = 512) -> List[Chunk]:
    chunks: List[Chunk] = []
    for section in act.sections:
        chunks.extend(chunk_section(act, section, token_threshold))
    return chunks

File: web/scripts/lib/chunker.ts
// Adaptive hierarchical chunking, ported from ingestion/chunker.py. A section
// under the token threshold is stored as one whole chunk; a larger section is
// split into per-sub-section child chunks that each carry a parentSectionId.
// Sub-section text is stored once (parent expansion is a query-time concern).
import {
  type AmendmentHistory,
  type Chunk,
  type ProvenanceRecord,
} from "../../convex/lib/models";
import { type ParsedAct, type Section } from "./parser";

// Cheap, monotonic token estimate (whitespace words).
export function estimateTokens(text: string): number {
  return text.split(/\s+/).filter(Boolean).length;
}

export function sectionId(actId: string, sectionNumber: string): string {
  return `${actId}-${sectionNumber}`;
}

function provenance(
  act: ParsedAct,
  section: Section,
  verbatim: string,
  subSection?: string,
): ProvenanceRecord {
  return {
    actName: act.actName,
    actYear: act.actYear,
    actType: act.actType,
    sourceUrl: act.sourceUrl,
    sourceHash: act.sourceHash,
    retrievalDate: act.retrievalDate,
    verbatimText: verbatim,
    sectionNumber: section.sectionNumber,
    subSection,
  };
}

function amendmentHistory(section: Section): AmendmentHistory {
  return {
    entries: [...section.amendments],
    noneRecorded: section.amendments.length === 0,
  };
}

export function chunkSection(
  act: ParsedAct,
  section: Section,
  tokenThreshold: number,
): Chunk[] {
  const parentId = sectionId(act.actId, section.sectionNumber);
  const whole = section.fullText;
  if (estimateTokens(whole) <= tokenThreshold || section.subSections.length === 0) {
    return [
      {
        chunkId: parentId,
        actId: act.actId,
        sectionNumber: section.sectionNumber,
        text: whole,
        provenance: provenance(act, section, whole),
        amendmentHistory: amendmentHistory(section),
        isDefinition: section.isDefinition,
        tokenEstimate: estimateTokens(whole),
      },
    ];
  }

  const children: Chunk[] = [];
  // A chunkId is the point's primary key in the store, so it must be unique.
  // Sub-section labels can repeat within one section (e.g. a definitions
  // section where many clauses carry an inner "(i)"); disambiguate a repeated
  // label with a positional suffix while the displayed subSection keeps the
  // real label.
  const seen = new Map<string, number>();
  for (const sub of section.subSections) {
    const count = (seen.get(sub.label) ?? 0) + 1;
    seen.set(sub.label, count);
    const uniqueLabel = count === 1 ? sub.label : `${sub.label}-${count}`;
    children.push({
      chunkId: `${parentId}-${uniqueLabel}`,
      actId: act.actId,
      sectionNumber: section.sectionNumber,
      subSection: sub.label,
      parentSectionId: parentId,
      text: sub.text,
      provenance: provenance(act, section, sub.text, sub.label),
      amendmentHistory: amendmentHistory(section),
      isDefinition: section.isDefinition,
      tokenEstimate: estimateTokens(sub.text),
    });
  }
  return children;
}

export function chunkAct(act: ParsedAct, tokenThreshold = 512): Chunk[] {
  return act.sections.flatMap((section) =>
    chunkSection(act, section, tokenThreshold),
  );
}
